File: renard2024ros/part-ii/ros-workspace/src/pypkg/pypkg/h.py
import rclpy

from rclpy.action import ActionClient
from rclpy.action.client import ClientGoalHandle, GoalStatus
from rclpy.node import Node
from typing import Optional

from my_interfaces.action import CountUntil

class CountAction(Node):
    """A node to that is a client for an action server.
    """

    def __init__(self):
        """
        """

        super().__init__("h")

        self.logger = self.get_logger()
        self.goalhandle: Optional[ClientGoalHandle] = None

        # Create client.
        #
        # Here, we create all client(s) that this node is responsible for
        # handling.
        self.client = ActionClient(self, CountUntil, "count")
        return

    def goal(self, target: int, step: float):
        """Send a goal to the server.
        """

        self.client.wait_for_server()
        goal = CountUntil.Goal()

        # Set the goal.
        #
        # This fills in the [`CountUntil.Goal`] with the appropriate arguments
        # besides the default values.
        goal.target = target
        goal.step = step

        self.client.send_goal_async(goal, feedback_callback=self.feedback).add_done_callback(self.response)
        return

    def response(self, future):
        """The callback from the action server responds to a goal request.
        """

        self.goalhandle = future.result()

        if self.goalhandle.accepted:
            self.logger.info("goal accepted...")
            self.goalhandle.get_result_async().add_done_callback(self.result)

            return

        self.logger.info("goal rejected...")
        return


    def cancel(self):
        self.logger.info("sending cancellation request...")
        self.goalhandle.cancel_goal_async()
    
    def feedback(self, message):
        """Process the feedback from the server.
        """

        current = message.feedback.current        
        self.logger.info(f"feedback: `current={current}`")

        # This is a test.
        #
        # Here, we are testing the cancel request mechanism by checking if the
        # current is greater than some number.
        if current >= 5:
            self.cancel()

    def result(self, future):
        """The callback when the action server is done.
        """

        match future.result().status:
            case GoalStatus.STATUS_SUCCEEDED:
                self.logger.info("the goal request was successful...")
            case GoalStatus.STATUS_ABORTED:
                self.logger.error("the goal request was aborted...")
            case GoalStatus.STATUS_CANCELED:
                self.logger.warn("the goal request was canceled...")

        self.logger.info(f"`reached={future.result().result.reached}`")
        return

def main(args=None):
    """The main function.
    """

    rclpy.init(args=args)

    node = CountAction()
    node.goal(100, 0.25)

    # Spin-up the node.
    #
    # This will start the node up, accordingly. This function call is also thread
    # blocking until a signal stop is sent to it, accordingly.
    rclpy.spin(node)

    # Shut this node down, gracefully.
    #
    # This will shut the node down from the ROS communication statck without
    # causing any major hiccups, assumingly.
    rclpy.shutdown()

File: renard2024ros/part-ii/ros-workspace/src/cpppkg/src/h.cpp
#include "my_interfaces/action/count_until.hpp"

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"

using CountUntil = my_interfaces::action::CountUntil;
using CountUntilGoalHandle = rclcpp_action::ClientGoalHandle<CountUntil>;

class CountClient : public rclcpp::Node {
public:
  CountClient() : Node("h") {
    this->client = rclcpp_action::create_client<CountUntil>(this, "count");
  }

  auto goal(int target, double step) -> void {
    // Wait for the server.
    //
    // If this client makes a goal request and the server is not online, then we
    // shall wait indefinitely until the server is online.
    this->client->wait_for_action_server();

    auto goal = CountUntil::Goal();
    goal.target = target;
    goal.step = step;

    // Set the callbacks.
    //
    // This is different compared to the Python-based API for ROS as we need to
    // register the callbacks before sending the request.
    auto options = rclcpp_action::Client<CountUntil>::SendGoalOptions();

    options.goal_response_callback =
        std::bind(&CountClient::response, this, std::placeholders::_1);

    options.result_callback =
        std::bind(&CountClient::result, this, std::placeholders::_1);

    options.feedback_callback =
        std::bind(&CountClient::feedback, this, std::placeholders::_1,
                  std::placeholders::_2);

    // Submit goal request.
    //
    // This is the request made to the server for the goal from this client,
    // accordingly.
    this->client->async_send_goal(goal, options);
  }

  auto feedback(const CountUntilGoalHandle::SharedPtr &handle,
                const std::shared_ptr<const CountUntil::Feedback> feedback)
      -> void {
    // Ignore the parameter.
    //
    // This is set to avoid any compiler warnings upon compiling this
    // translation file, accordingly
    (void)handle;

    RCLCPP_INFO(this->get_logger(), "feedback: `counter=%ld`",
                feedback->current);

    if (feedback->current >= 5) {
      this->cancel();
    }
  }

  auto cancel() -> void {
    RCLCPP_INFO(this->get_logger(), "sending cancellation request...");
    this->client->async_cancel_all_goals();
  }

  // The callback for the response.
  //
  // This is the callback that gets called when the server responds to this
  // client's goal request.
  auto response(const CountUntilGoalHandle::SharedPtr &handle) -> void {
    if (!handle) {
      RCLCPP_INFO(this->get_logger(), "goal rejected...");
      return;
    }

    RCLCPP_INFO(this->get_logger(), "goal accepted...");
  }

  auto result(const CountUntilGoalHandle::WrappedResult &result) -> void {
    auto status = result.code;

    if (status == rclcpp_action::ResultCode::SUCCEEDED) {
      // The server processed the goal.
      //
      // This means that the goal request sent by this client was successfully
      // completed and returned.
      RCLCPP_INFO(this->get_logger(), "the goal request was successful...");
    } else if (status == rclcpp_action::ResultCode::ABORTED) {
      // The server terminated early.
      //
      // This can be due to multiple reasons. However, it should be known that
      // the server originally accepted the request but found an issues that
      // prevented the goal from being reached.
      RCLCPP_WARN(this->get_logger(), "the goal request was aborted...");
    } else if (status == rclcpp_action::ResultCode::CANCELED) {
      // The server goal request was canceled.
      //
      // This indicates that the goal was sent, accepted, and proceeded.
      // However, the client sent a request that was accepted to cancel the
      // current in-progress goal.
      RCLCPP_INFO(this->get_logger(), "the goal request was canceled...");
    }

    // Print the result.
    //
    // This is a temporary next-step that allows us to debug and ensure that the
    // goal was indeed met, accordingly.
    RCLCPP_INFO(this->get_logger(), "`reached=%ld`", result.result->reached);
  }

  rclcpp_action::Client<CountUntil>::SharedPtr client;
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);

  auto node = std::make_shared<CountClient>();
  node->goal(10, 0.5);

  // Spin-up the ROS-based node.
  //
  // This will run the ROS-styled node infinitely until the signal to stop the
  // program is received, accordingly.
  rclcpp::spin(node);

  // Shut the node down, gracefully.
  //
  // This will close and exit the node execution without disrupting the ROS
  // communication network, assumingly.
  rclcpp::shutdown();

  // The final return.
  //
  // This is required for the main function of a program within the C++
  // programming language.
  return 0;
}
